third_party/txt/src/txt/paragraph.cc

```diff
@@ -1483,6 +1589,51 @@ Paragraph::PositionWithAffinity Paragraph::GetGlyphPositionAtCoordinate(
   }
 }
 
+// We don't cache this because since this returns all boxes, it is usually
+// unnecessary to call this multiple times in succession.
+std::vector<Paragraph::TextBox> Paragraph::GetRectsForWidgets() const {
+  // Struct that holds calculated metrics for each line.
+  struct LineBoxMetrics {
+    std::vector<Paragraph::TextBox> boxes;
+    // Per-line metrics for max and min coordinates for left and right boxes.
+    // These metrics cannot be calculated in layout generically because of
+    // selections that do not cover the whole line.
+    SkScalar max_right = FLT_MIN;
+    SkScalar min_left = FLT_MAX;
+  };
+
+  std::map<size_t, LineBoxMetrics> line_metrics;
+
+  // Generate initial boxes and calculate metrics.
+  for (const CodeUnitRun& run : inline_widget_code_unit_runs_) {
+    // Check to see if we are finished.
+    double baseline = line_baselines_[run.line_number];
+    SkScalar top = baseline + run.font_metrics.fAscent;
+    SkScalar bottom = baseline + run.font_metrics.fDescent;
+
+    if (run.widget_run != nullptr) {  // Use inline widget size as height.
+      top = baseline - run.widget_run->baseline;
+      bottom = baseline + run.widget_run->height - run.widget_run->baseline;
+    }
+
+    // Calculate left and right.
+    SkScalar left, right;
+    left = run.x_pos.start;
+    right = run.x_pos.end;
+
+    line_metrics[run.line_number].boxes.emplace_back(
```

Comment:
Is it necessary to group the boxes by line number?  Or can they be inserted into the result list here?